hal/watchdog_api.h

```diff
@@ -0,0 +1,116 @@
+/** \addtogroup hal */
+/** @{*/
+/* mbed Microcontroller Library
+ * Copyright (c) 2017 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef MBED_WATCHDOG_API_H
+#define MBED_WATCHDOG_API_H
+
+#include <stdbool.h>
+#include <stdint.h>
+
+typedef struct
+{
+  // Refresh value for the watchdog in milliseconds. The maximum value of this
+  // setting is platform dependent, to find the maximum value for the current
+  // platform call hal_watchdog_get_max_timeout(void)
+  uint32_t timeout_ms;
+  // Enables the window watchdog. This will cause the API to run the window
+  // watchdog instead of the independent watchdog if supported by the current
+  // platform. This is false by default.
+  bool enable_window;
+  // Window watchdog configuration value in milliseconds. This defines the
+  // earliest time period (in milliseconds) after refreshing the watchdog in
+  // which the watchdog can be refreshed again. This value must be smaller than
+  // the timeout_ms setting.
+  uint32_t window_ms;
+  // Enables or disables the watchdog in sleep mode. By default when the system
+  // is in sleep mode the watchdog is paused. When this flag is enabled the
+  // watchdog will continue to run while the system is asleep.
+  // This is disabled by default.
+  bool enable_sleep;
+} watchdog_config_t;
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+// Initialise and start a watchdog timer with the given configuration
+//
+// This will initialise the watchdog timer with the given configuration settings
+// if any specific settings are not supported by the current platform the
+// function will assert.
+//
+// The watchdog timer will start at the end of this function call and will not
+// be stopped until the system is reset, or if supported the stop function is
+// called.
+//
+// @warning This function will assert if the enabled parameters are not
```

Comment:
I'd suggest an hard `error` instead of a call to `assert`.

Reply:
As Martin suggested I will add a watchdog error type to handle this failing instead of terminating the program.